cargo/crate_universe_resolver/src/main.rs

```diff
@@ -0,0 +1,111 @@
+use anyhow::{anyhow, Context};
+use log::*;
+use crate_universe_resolver::config::Config;
+use std::io::{BufRead, BufReader};
+use std::path::PathBuf;
+use structopt::StructOpt;
+
+// Options which don't affect the contents of the generated should be on this struct.
+// These fields are not factored into cache keys.
+//
+// Anything which affects the contents of the generated output should live on `config::Config`.
+#[derive(StructOpt)]
+struct Opt {
+    #[structopt(long)]
+    repo_name: String,
+    #[structopt(long = "input_path", parse(from_os_str))]
+    input_path: PathBuf,
+    #[structopt(long = "output_path", parse(from_os_str))]
+    output_path: PathBuf,
+    #[structopt(long = "lockfile", parse(from_os_str))]
+    lockfile: Option<PathBuf>,
+    #[structopt(long = "update-lockfile")]
+    update_lockfile: bool,
+}
+
+fn main() -> anyhow::Result<()> {
+    env_logger::init();
+
+    let opt = Opt::from_args();
+
+    trace!("Parsing config from {:?}", opt.input_path);
+
+    let config: Config = {
+        let config_file = std::fs::File::open(&opt.input_path)
+            .with_context(|| format!("Failed to open config file at {:?}", opt.input_path))?;
+        serde_json::from_reader(config_file)
+            .with_context(|| format!("Failed to parse config file {:?}", opt.input_path))?
+    };
+    let lockfile = opt.lockfile;
+    let repo_name = opt.repo_name;
+    let output_path = opt.output_path;
+
+    trace!("Preprocessing config");
+    let mut resolver = config.preprocess()?;
+
+    if opt.update_lockfile {
+        if lockfile.is_none() {
+            eprintln!("Not updating lockfile for `crate_universe` repository with name \"{}\" because it has no `lockfile` attribute.", repo_name);
+        }
+    } else if let Some(lockfile) = lockfile {
+        let mut lockfile_format_version_line = String::new();
+        let mut lockfile_hash_line = String::new();
+        {
+            std::fs::File::open(&lockfile)
+                .map(BufReader::new)
+                .and_then(|mut f| {
+                    f.read_line(&mut lockfile_format_version_line)?;
+                    f.read_line(&mut lockfile_hash_line)?;
+                    Ok(())
+                })
+                .with_context(|| format!("Failed to read lockfile header from {:?}", lockfile))?;
+        }
+        if lockfile_format_version_line != "# rules_rust crate_universe file format 1\n" {
+            return Err(anyhow!("Unrecognized lockfile format"));
+        }
+        if let Some(lockfile_hash) = lockfile_hash_line.strip_prefix("# config hash ") {
```

Comment:
My biggest gripe with the current implementation is the "lockfiles" and the need for this hash. If the rule was not marked as `local` then it wouldn't trigger every build and could rely on just the `Cargo.toml` tree and `Cargo.lock` files. Typically (at least in my experience with `cargo-raze`) regenerating BUILD files is not terribly expensive and is pretty quick. I think it'd be a lot simpler and effective to allow Bazel to run the rule when it believes it should be (not using `local`) and allow users to `bazel sync` when they need new dependencies. Typically though, if the `Cargo.lock` file is a field on the rule then all they would need to do is update that file.

Reply:
> My biggest gripe with the current implementation is the "lockfiles" and the need for this hash.

> If the rule was not marked as `local` then it wouldn't trigger every build and could rely on just the `Cargo.toml` tree and `Cargo.lock` files. Typically (at least in my experience with `cargo-raze`) regenerating BUILD files is not terribly expensive and is pretty quick.

I have regularly seen resolves take double-digit seconds, which is far too much time to do every time you `git pull` a new change in.

> I think it'd be a lot simpler and effective to allow Bazel to run the rule when it believes it should be (not using `local`) and allow users to `bazel sync` when they need new dependencies. Typically though, if the `Cargo.lock` file is a field on the rule then all they would need to do is update that file.

I agree with removing the `local`, but a few really important UX points when using a lockfile are:
1. If I `git pull` and do a build, I get the same versions of dependencies as the person who committed without needing to do a resolve.
2. If I change any of the input to the rule, I don't silently get the old versions.
3. There's an easy way for me to say "I want new versions, subject to the same version constraints".

I think making the lockfile actually be a `Cargo.lock` file would be fine _if it contains all of the information needed to guarantee the above properties_, and would be happy to review (or work on) a patch to make that happen. But I'd rather not block on that re-work, if we can avoid it.

Reply:
> If I git pull and do a build, I get the same versions of dependencies as the person who committed without needing to do a resolve.

This sounds like magic to me. You'd have to be able to trust the "lockfile" that's there and simply not do the resolve. But you can't know that the lockfile is up to date without first resolving it yourself. Right?

> If I change any of the input to the rule, I don't silently get the old versions.

I don't see how this would be the case if all inputs were tracked. But this also assumes the rule would resolve each time to ensure all dependencies were tracked

> There's an easy way for me to say "I want new versions, subject to the same version constraints".

Is this analogous to `cargo update`? 

> I have regularly seen resolves take double-digit seconds, which is far too much time to do every time you `git pull` a new change in.

I think because the rule was marked as [local](https://docs.bazel.build/versions/master/skylark/repository_rules.html#when-is-the-implementation-function-executed) and doesn't return any repository info (see [http.bzl](https://github.com/bazelbuild/bazel/blob/master/tools/build_defs/repo/http.bzl#L123)) it'd always retrigger.

Reply:
> > If I git pull and do a build, I get the same versions of dependencies as the person who committed without needing to do a resolve.
> 
> This sounds like magic to me. You'd have to be able to trust the "lockfile" that's there and simply not do the resolve. But you can't know that the lockfile is up to date without first resolving it yourself. Right?

This is what the rules offer today, if you use the current lockfile mechanism. We hash every input into the resolution process (including the resolver binary itself, the cargo binary, and all arguments to everything), and check whether they're up-to-date.

> > There's an easy way for me to say "I want new versions, subject to the same version constraints".
> 
> Is this analogous to `cargo update`?

Yes

> > I have regularly seen resolves take double-digit seconds, which is far too much time to do every time you `git pull` a new change in.
> 
> I think because the rule was marked as [local](https://docs.bazel.build/versions/master/skylark/repository_rules.html#when-is-the-implementation-function-executed) and doesn't return any repository info (see [http.bzl](https://github.com/bazelbuild/bazel/blob/master/tools/build_defs/repo/http.bzl#L123)) it'd always retrigger.

I don't mean "I see it doing double-digit second resolves every build", I mean "A single resolve can take double-digit seconds" - I'm not saying "it triggers too often", I'm saying "we should minimise triggering, because any single trigger can be very expensive", but I am totally cool with achieving this by fixing up how the rule runs (e.g. by properly setting up params). I do, however, think the "If I git pull, and am using a lockfile, I don't need to pay for a resolve" is really important.

Reply:
>   I don't mean "I see it doing double-digit second resolves every build", I mean "A single resolve can take double-digit seconds" - I'm not saying "it triggers too often", I'm saying "we should minimise triggering, because any single trigger can be very expensive", but I am totally cool with achieving this by fixing up how the rule runs (e.g. by properly setting up params). I do, however, think the "If I git pull, and am using a lockfile, I don't need to pay for a resolve" is really important.

I 100% agree. The thing I care most about is being able to take an existing Cargo files and slot them into this Rule as the source of truth without requiring anything additional to be generated. Though, I like the idea that users can specify a hash value to indicate whether or not resolution needs to be done (if that's what you mean by)

> This is what the rules offer today, if you use the current lockfile mechanism. We hash every input into the resolution process (including the resolver binary itself, the cargo binary, and all arguments to everything), and check whether they're up-to-date.